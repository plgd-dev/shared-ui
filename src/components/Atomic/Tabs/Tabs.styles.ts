import { css } from '@emotion/react'
import { colors } from '../_utils/colors'
import { fontPrimary } from '../_utils/commonStyles'
import { ThemeType, get } from '../_theme'

export const container = css`
    overflow-y: hidden;
    box-shadow: none;
`

export const fullHeight = css`
    flex: 1 1 auto;
    display: flex;
    flex-direction: column;
    height: 100%;
`

export const tabList = (theme: ThemeType) => css`
    display: block;
    position: relative;
    border-bottom: 1px solid ${get(theme, `Tabs.list.borderColor`)};
`

export const tabListInnerPadding = css`
    margin-left: 40px;
    margin-right: 40px;
`

export const tabItem = (theme: ThemeType) => css`
    white-space: nowrap;
    box-sizing: border-box;
    text-align: center;
    text-rendering: optimizelegibility;
    user-select: none;
    outline: none;
    box-shadow: none;
    cursor: pointer;
    text-decoration: none;
    border-width: initial;
    border-style: none;
    border-color: initial;
    border-image: initial;
    padding: 16px 0;
    text-size-adjust: none;
    text-overflow: ellipsis;
    font-family: ${fontPrimary};
    font-style: normal;
    font-weight: 600;
    font-size: 14px;
    line-height: 16px;
    color: ${colors.neutral500};
    overflow: hidden;
    background: transparent;
    margin: 0 24px 0 0;
    transition: all 0.25s;
    display: inline-flex;
    align-items: center;

    &:hover {
        color: ${get(theme, 'Tabs.item.hover.color')};
        background: transparent;
    }

    &:last-child {
        margin-right: 0;
    }
`

export const isActive = (theme: ThemeType) => css`
    color: ${get(theme, `Tabs.active.color`)};
`

export const isDisabled = (theme: ThemeType) => css`
    color: ${get(theme, `Tabs.disabled.color`)};
    cursor: not-allowed !important;

    &:hover {
        color: ${get(theme, `Tabs.disabled.color`)};
    }
`

export const slider = (theme: ThemeType) => css`
    height: 3px;
    bottom: 0;
    position: absolute;
    background: ${get(theme, `Tabs.slider.background`)};
`

export const page = css`
    padding-top: 24px;
`

export const pageXpadding = css`
    padding-left: 40px;
    padding-right: 40px;
    overflow: auto;
`

export const icon = css`
    margin-left: 8px;
`

export const iconSuccess = (theme: ThemeType) => css`
    color: ${get(theme, `Tabs.icon.success.color`)};
`

export const iconError = (theme: ThemeType) => css`
    color: ${get(theme, `Tabs.icon.error.color`)};
`
