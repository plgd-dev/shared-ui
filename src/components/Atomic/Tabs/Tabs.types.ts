import { ReactNode } from 'react'

export type TabItem = {
    id: number
    name: string
    content: ReactNode
}

export type Props = {
    activeItem?: number
    fullHeight?: boolean
    onItemChange?: (activeItem: number) => void
    tabs: TabItem[]
}

export const defaultProps = {
    activeItem: 0,
}
