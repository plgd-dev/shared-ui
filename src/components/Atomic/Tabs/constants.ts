export const statuses = {
    SUCCESS: 'success',
    ERROR: 'error',
} as const
