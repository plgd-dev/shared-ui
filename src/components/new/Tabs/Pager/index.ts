export { default } from './Pager'
export * from './Pager'
