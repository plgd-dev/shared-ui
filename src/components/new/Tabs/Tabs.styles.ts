import { css } from '@emotion/react'
import { colors } from '../_utils/colors'
import { fontPrimary } from '../_utils/commonStyles'

export const container = css`
    overflow-y: hidden;
    box-shadow: none;
`

export const tabList = css`
    display: block;
    position: relative;
    border-bottom: 1px solid ${colors.neutral200};
`

export const tabItem = css`
    white-space: nowrap;
    box-sizing: border-box;
    text-align: center;
    text-rendering: optimizelegibility;
    user-select: none;
    outline: none;
    box-shadow: none;
    cursor: pointer;
    text-decoration: none;
    border-width: initial;
    border-style: none;
    border-color: initial;
    border-image: initial;
    padding: 16px 0;
    text-size-adjust: none;
    text-overflow: ellipsis;
    font-family: ${fontPrimary};
    font-style: normal;
    font-weight: 600;
    font-size: 14px;
    line-height: 16px;
    color: ${colors.neutral500};
    overflow: hidden;
    background: transparent;
    margin: 0 24px 0 0;
    transition: all 0.25s;

    &:hover {
        color: ${colors.primaryDarken};
        background: transparent;
    }

    &:last-child {
        margin-right: 0;
    }
`

export const isActive = css`
    color: ${colors.primaryDarken};
`

export const slider = css`
    height: 3px;
    bottom: 0;
    position: absolute;
    background: ${colors.primaryBonus};
`

export const page = css`
    padding: 24px 0 0 0;
`
