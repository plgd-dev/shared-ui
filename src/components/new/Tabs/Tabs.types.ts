import { ReactNode } from 'react'

export type TabItem = {
    name: string
    content: ReactNode
}

export type Props = {
    tabs: TabItem[]
    activeItem?: number
    onItemChange?: (activeItem: number) => void
}

export const defaultProps = {
    activeItem: 0,
}
