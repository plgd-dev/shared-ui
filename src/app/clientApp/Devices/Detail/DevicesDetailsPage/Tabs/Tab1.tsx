import { FC, useEffect, useMemo, useState } from 'react'
import { useIntl } from 'react-intl'

import SimpleStripTable from '../../../../../../components/Atomic/SimpleStripTable'
import TagGroup, { justifyContent } from '../../../../../../components/Atomic/TagGroup'
import Tag from '../../../../../../components/Atomic/Tag'
import { DEVICE_PROVISION_STATUS_DELAY_MS } from '../../../constants'
import { IconLoader } from '../../../../../../components/Atomic/Loader'
import { getColorByOnboardingStatus, getColorByProvisionStatus, getDPSEndpoint, loadResourceData } from '../../../utils'
import { messages as t } from '../../../Devices.i18n'
import { messages as app } from '../../../../App/App.i18n'
import { Props } from './Tab1.types'
import testId from '../../../../testId'
import StatusTag from '../../../../../../components/Atomic/StatusTag/StatusTag'
import { TagTypeType } from '../../../../../../components/Atomic/StatusTag/StatusTag.types'

const Tab1: FC<Props> = (props) => {
    const { data, deviceId, deviceOnboardingResourceData, isOwned, isUnsupported, resources, onboardResourceLoading } = props
    const { formatMessage: _ } = useIntl()

    const { onboardTitleStatus } = testId.devices.detail

    const [resourceLoading, setResourceLoading] = useState(false)
    const [deviceResourceData, setDeviceResourceData] = useState<any>(undefined)

    const provisionStatus = deviceResourceData?.content?.provisionStatus
    const onboardingStatus = deviceOnboardingResourceData?.content?.cps || 'n/a'
    const dpsEndpoint = useMemo(() => getDPSEndpoint(resources), [resources])

    useEffect(() => {
        if (dpsEndpoint && isOwned && !deviceResourceData) {
            setResourceLoading(true)
            setTimeout(() => {
                loadResourceData({
                    href: dpsEndpoint.href,
                    deviceId,
                    errorCallback: () => {
                        setResourceLoading(false)
                    },
                }).then((rData) => {
                    setDeviceResourceData(rData)
                    setResourceLoading(false)
                })
            }, DEVICE_PROVISION_STATUS_DELAY_MS)
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [dpsEndpoint, isOwned, deviceResourceData])

    const rows = [
        { attribute: _(t.id), value: deviceId },
        {
            attribute: _(t.types),
            value: data?.types ? (
                <TagGroup
                    i18n={{
                        more: _(app.more),
                        modalHeadline: _(app.types),
                    }}
                    justifyContent={justifyContent.END}
                >
                    {data?.types.map((t, key) => <Tag key={key}>{t}</Tag>)}
                </TagGroup>
            ) : (
                <div>-</div>
            ),
        },
        {
            attribute: _(t.ownershipStatus),
            value: <StatusTag variant={isOwned ? 'success' : 'error'}>{isOwned ? _(t.owned) : _(t.unowned)}</StatusTag>,
            hidden: isUnsupported,
        },
        {
            attribute: _(t.onboardingStatus),
            value: onboardResourceLoading ? (
                <IconLoader size={20} type='secondary' />
            ) : (
                <StatusTag data-test-id={onboardTitleStatus} variant={getColorByOnboardingStatus(onboardingStatus) as TagTypeType}>
                    {onboardingStatus}
                </StatusTag>
            ),
        },
        {
            attribute: _(t.dpsStatus),
            value: resourceLoading ? (
                <IconLoader size={20} type='secondary' />
            ) : (
                <StatusTag variant={isOwned ? (getColorByProvisionStatus(provisionStatus) as TagTypeType) : 'normal'}>
                    {isOwned && deviceResourceData ? provisionStatus : _(t.notAvailable)}
                </StatusTag>
            ),
            hidden: !dpsEndpoint,
        },
        {
            attribute: _(t.endpoints),
            value: data?.endpoints ? (
                <TagGroup
                    i18n={{
                        more: _(app.more),
                        modalHeadline: _(app.types),
                    }}
                    justifyContent={justifyContent.END}
                >
                    {data?.endpoints?.map?.((endpoint: string) => <Tag key={endpoint}>{endpoint}</Tag>)}
                </TagGroup>
            ) : (
                <div>-</div>
            ),
        },
    ]

    return (
        <div
            style={{
                paddingTop: 8,
                overflow: 'hidden',
            }}
        >
            <SimpleStripTable rows={rows.filter((r) => r?.hidden === false || r?.hidden === undefined)} />
        </div>
    )
}

Tab1.displayName = 'Tab1'

export default Tab1
