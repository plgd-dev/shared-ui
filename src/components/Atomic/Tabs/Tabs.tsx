import { FC, useEffect, useRef, useState } from 'react'
import isFunction from 'lodash/isFunction'
import { motion } from 'framer-motion'

import { defaultProps, Props } from './Tabs.types'
import * as styles from './Tabs.styles'
import { useMeasure } from '../../../common/hooks/use-measure'
import Pager from './Pager'

const Tabs: FC<Props> = (props) => {
    const { activeItem, onItemChange, fullHeight, tabs } = { ...defaultProps, ...props }
    const [value, setValue] = useState(activeItem)
    const childRefs = useRef(new Map())
    const tabListRef = useRef<HTMLDivElement | null>(null)
    const [slider, setSlider] = useState({ left: 0, right: 0, hasValue: false })
    const { bounds, ref } = useMeasure()

    const handleTabChange = (i: number) => {
        setValue(i)
        isFunction(onItemChange) && onItemChange(i)
    }

    // measure our elements
    useEffect(() => {
        const target = childRefs.current.get(value)
        const container = tabListRef.current
        if (target && container) {
            const cRect = container.getBoundingClientRect()

            // when container is `display: none`, width === 0.
            // ignore this case
            if (cRect.width === 0) {
                return
            }

            const tRect = target.getBoundingClientRect()
            const left = tRect.left - cRect.left
            const right = cRect.right - tRect.right

            setSlider({
                hasValue: true,
                left,
                right,
            })
        }
    }, [value, bounds])

    return (
        <div css={[styles.container, fullHeight && styles.fullHeight]} ref={ref}>
            <div css={styles.tabList} ref={tabListRef}>
                {tabs.map((tab, i) => (
                    <motion.button
                        css={[styles.tabItem, i === value && styles.isActive]}
                        key={i}
                        onClick={() => handleTabChange(tab.id)}
                        ref={(el) => childRefs.current.set(i, el)}
                        transition={{ duration: 0.25 }}
                    >
                        {tab.name}
                    </motion.button>
                ))}
                {slider.hasValue && (
                    <motion.div
                        layout
                        css={styles.slider}
                        initial={false}
                        style={{
                            left: slider.left,
                            right: slider.right,
                        }}
                        transition={{ bounceDamping: 3 }}
                    />
                )}
            </div>
            <Pager fullHeight={fullHeight} value={value}>
                {tabs.map((tab, i) => (
                    <div
                        css={styles.page}
                        key={i}
                        style={{
                            width: '100%',
                            height: fullHeight ? '100%' : undefined,
                        }}
                    >
                        {tab.content}
                    </div>
                ))}
            </Pager>
        </div>
    )
}

Tabs.displayName = 'Tabs'
Tabs.defaultProps = defaultProps

export default Tabs
