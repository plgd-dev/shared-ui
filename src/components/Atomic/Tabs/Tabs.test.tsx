import React from 'react'
import { render } from '@testing-library/react'
import Tabs from './Tabs'

describe('<Tabs>', () => {
    it('render correctly - snapshot', () => {
        const { asFragment } = render(
            <div>
                <Tabs
                    onItemChange={(activeItem) => console.log(`Active item: ${activeItem}`)}
                    tabs={[
                        { name: 'Device information', id: 0, content: <div style={{ height: 300, background: '#e5e5e5' }}>Device information content</div> },
                        { name: 'Resources', id: 1, content: <div style={{ height: 300, background: '#e5e5e5' }}>Resources content</div> },
                    ]}
                />
            </div>
        )

        expect(asFragment()).toMatchSnapshot()
    })
})
