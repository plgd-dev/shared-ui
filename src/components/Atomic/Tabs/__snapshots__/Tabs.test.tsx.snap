// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs> render correctly - snapshot 1`] = `
<DocumentFragment>
  .emotion-0 {
  overflow-y: hidden;
  box-shadow: none;
}

.emotion-1 {
  display: block;
  position: relative;
  border-bottom: 1px solid;
}

.emotion-2 {
  white-space: nowrap;
  box-sizing: border-box;
  text-align: center;
  text-rendering: optimizelegibility;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  outline: none;
  box-shadow: none;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  border-width: initial;
  border-style: none;
  border-color: initial;
  border-image: initial;
  padding: 16px 0;
  -webkit-text-size-adjust: none;
  -moz-text-size-adjust: none;
  -ms-text-size-adjust: none;
  text-size-adjust: none;
  text-overflow: ellipsis;
  font-family: 'Poppins',sans-serif;
  font-style: normal;
  font-weight: 600;
  font-size: 14px;
  line-height: 16px;
  color: #81868C;
  overflow: hidden;
  background: transparent;
  margin: 0 24px 0 0;
  -webkit-transition: all 0.25s;
  transition: all 0.25s;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-2:hover {
  background: transparent;
}

.emotion-2:last-child {
  margin-right: 0;
}

.emotion-3 {
  white-space: nowrap;
  box-sizing: border-box;
  text-align: center;
  text-rendering: optimizelegibility;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  outline: none;
  box-shadow: none;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  border-width: initial;
  border-style: none;
  border-color: initial;
  border-image: initial;
  padding: 16px 0;
  -webkit-text-size-adjust: none;
  -moz-text-size-adjust: none;
  -ms-text-size-adjust: none;
  text-size-adjust: none;
  text-overflow: ellipsis;
  font-family: 'Poppins',sans-serif;
  font-style: normal;
  font-weight: 600;
  font-size: 14px;
  line-height: 16px;
  color: #81868C;
  overflow: hidden;
  background: transparent;
  margin: 0 24px 0 0;
  -webkit-transition: all 0.25s;
  transition: all 0.25s;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-3:hover {
  background: transparent;
}

.emotion-3:last-child {
  margin-right: 0;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: hidden;
  width: 100%;
}

.emotion-5 {
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  direction: ltr;
  will-change: transform;
  min-height: 0;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  height: 100%;
  overflow-y: hidden;
  outline: none;
}

.emotion-7 {
  padding-top: 24px;
}

<div>
    <div
      class="emotion-0"
    >
      <div
        class="emotion-1"
      >
        <button
          class="emotion-2"
        >
          Device information
        </button>
        <button
          class="emotion-3"
        >
          Resources
        </button>
      </div>
      <div
        class="emotion-4"
      >
        <div
          class="emotion-5"
          style="transform: translateX(0%) translateZ(0);"
        >
          <div
            aria-hidden="false"
            class="emotion-6"
            tabindex="0"
          >
            <div
              class="emotion-7"
              style="width: 100%;"
            >
              <div
                style="height: 300px; background: rgb(229, 229, 229);"
              >
                Device information content
              </div>
            </div>
          </div>
          <div
            aria-hidden="true"
            class="emotion-6"
            tabindex="-1"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
